pkg/webhooks/kafkacluster_validator.go

```diff
@@ -133,3 +157,65 @@ func getMissingMounthPathLocation(mounthPath string, kafkaClusterSpec *banzaiclo
 }
+
+// checkListeners validates the spec.listenersConfig object
+func checkInternalAndExternalListeners(kafkaClusterSpec *banzaicloudv1beta1.KafkaClusterSpec) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, checkUniqueListenerContainerPort(kafkaClusterSpec.ListenersConfig)...)
+
+	allErrs = append(allErrs, checkExternalListenerStartingPort(kafkaClusterSpec)...)
+
+	return allErrs
+}
+
+// checkUniqueListenerContainerPort checks for duplicate containerPort numbers across both internal and external listeners
+// which would subsequently generate a "Duplicate value" error when creating a Service which accumulates all these ports.
+// The first time a port number is found will not be reported as duplicate; only subsequent instances using that port are.
+// (this is done to keep in tune with the way the K8s Service would report the "Duplicate value" error which ignores the first instance)
+func checkUniqueListenerContainerPort(listeners banzaicloudv1beta1.ListenersConfig) field.ErrorList {
+	var allErrs field.ErrorList
+	var containerPorts = make(map[int32]int)
+
+	for i, intListener := range listeners.InternalListeners {
+		containerPorts[intListener.ContainerPort] += 1
+		if containerPorts[intListener.ContainerPort] > 1 {
+			fldErr := field.Duplicate(field.NewPath("spec").Child("listenersConfig").Child("internalListeners").Index(i).Child("containerPort"), intListener.ContainerPort)
+			allErrs = append(allErrs, fldErr)
+		}
+	}
+	for i, extListener := range listeners.ExternalListeners {
+		containerPorts[extListener.ContainerPort] += 1
+		if containerPorts[extListener.ContainerPort] > 1 {
+			fldErr := field.Duplicate(field.NewPath("spec").Child("listenersConfig").Child("externalListeners").Index(i).Child("containerPort"), extListener.ContainerPort)
+			allErrs = append(allErrs, fldErr)
+		}
+	}
+
+	return allErrs
+}
+
+// checkExternalListenerStartingPort checks the generic sanity of the resulting external port (valid number between 1 and 65535)
+func checkExternalListenerStartingPort(kafkaClusterSpec *banzaicloudv1beta1.KafkaClusterSpec) field.ErrorList {
+	// if there are no externalListeners, there is no need to perform the rest of the checks in this function
+	if kafkaClusterSpec.ListenersConfig.ExternalListeners == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	const maxPort int32 = 65535
+	for i, extListener := range kafkaClusterSpec.ListenersConfig.ExternalListeners {
+		var invalidBrokerIDs []int32
+		for _, broker := range kafkaClusterSpec.Brokers {
+			if extListener.ExternalStartingPort+broker.Id < 1 || extListener.ExternalStartingPort+broker.Id > maxPort {
+				invalidBrokerIDs = append(invalidBrokerIDs, broker.Id)
+			}
+		}
+		if len(invalidBrokerIDs) > 0 {
+			errmsg := invalidExternalListenerStartingPortErrMsg + ": " + fmt.Sprintf("ExternalListener '%s' would generate invalid port numbers (not between 1 and 65535) for brokers %v", extListener.Name, invalidBrokerIDs)
```

Comment:
I think we should chage a bit on this error message ""ExternalListener '%s' would generate invalid port numbers (not between 1 and 65535) for brokers %v"
Like: for externalListener '%s' the generated port numbers for external access (externalStartingPort '%d' + brokerID '%v') is not between 1 and 65535"

Reply:
For reference, currently, the message in question looks like this:
```
The KafkaCluster "kafka" is invalid:
* spec.listenersConfig.externalListeners[0].externalStartingPort: Invalid value: 65535: invalid external listener starting port number: ExternalListener 'external1' would generate invalid port numbers (not between 1 and 65535) for brokers [1 200]
* spec.listenersConfig.externalListeners[1].externalStartingPort: Invalid value: 65400: invalid external listener starting port number: ExternalListener 'external2' would generate invalid port numbers (not between 1 and 65535) for brokers [200]
```

How do you feel about updating **the last part** to look like this ?   
>ExternalListener 'external1' would generate external access port numbers (externalStartingPort 65535 + Broker IDs [11 102]) that are out of range (not between 1 and 65535)



A couple of notes:
1. I like the idea of inserting the "formula" in the log because that makes it significantly clearer for the user what the problem is
2. Having the problematic externalStartingPort and broker IDs right next to each other in the log could make it easier for our users to interpret the message
3. I favour the use of the conditional "would" in a validating webhook context because using past tense ("generated") gives me, as a user, the feeling that something already happened in the background (objects were created, who knows if they were cleaned up...) in the cluster which is both disconcerting and not true in this case.
4. I don't like "brokerIDs" as part of the string because it looks like a field name (and it isn't ;  for us the actual path would be `brokers[].id`). That's why I'd go with either `brokers.id %v` or `Broker IDs %v` to not give the impression that we're referencing a hidden/unknown field from the spec.

Reply:
IMO the current error message is very clear

Reply:
We should put somehow the formula. I think it is not so clear for the user. Most of the cases the users will not know how it is generated. Also it is hard to find in our documentation. I just like to see the formula with or without the values. 
Agree with the 3 and 4

Reply:
Here are a couple of examples that I think follow those thoughts :

1. formula added as an explanation and numbers are right in it including repeating the externalStartingPort :
    * `ExternalListener 'external1' would generate external access port numbers (externalStartingPort 65535 + Broker IDs [11 102]) that are out of range (not between 1 and 65535)`
        * this keeps relevant pieces of information (those user inputs) closer to each other 

2. formula added as an explanation but without the numbers right in it and without repeating the externalStartingPort value :
    * `ExternalListener 'external1' would generate external access port numbers (externalStartingPort + Broker ID) that are out of range (not between 1 and 65535) for brokers [11 102]`

What do you think about these ?

Reply:
I vote for  2. :)

Reply:
Done in commit c6f0ea75642551310b64e87faa4969e6c89e5f3b.

I also updated the PR description with the new log message (taken from a real install running this code against the same input as before).

Reply:
> We should show somehow the formula. I think it is not so clear for the user. Most of the cases the users will not know how is it generated. Also it is hard to find in our documentation. I would like to see the formula with or without the values.

Then we should show how this is calculated in the CRD level, not just the error log message

Reply:
👎🏼  It means I dont agree with that comment and because you also said that "feel free to merge"  before it has not been completely discussed

Reply:
> 👎🏼 It means I dont agree with that comment and because you also said that "feel free to merge" before it has not been completely discussed

"feel free to merge" it's purely from my own perspective, it's 100% ok to disagree, but please provide explanation instead of just a 👎🏼 , what if I give your 👎🏼 every time I don't agree with your opinion?

Reply:
On top of that, you also mentioned "LGTM", so your 👎🏼 is conflicting your own message to me

Reply:
I agree we should add the formula into the CRD. That should be in a separate PR because It is belongs to the API module.

Reply:
> I agree we should add the formula into the CRD. That should be in a separate PR because It is belongs to the API module.

I don't necessary agree we should do it in a separate PR because we are literally changing the API here in this PR, IMO we should just add a couple of lines here in this PR. But I will leave the decision to you @mihaialexandrescu

Reply:
I am trying to understand what's necessary and to come up with a proposal.   

So here goes my first try :  

In `api/v1beta1/kafkacluster_types.go` we would have : 
```
// ExternalListenerConfig defines the external listener config for Kafka
type ExternalListenerConfig struct {
	... ...
	// +kubebuilder:validation:Minimum=0
	// +kubebuilder:validation:Maximum=65535
	// externalStartingPort is added to each broker ID to get the port number that will be used for external access to the broker. 
	// If accessMethod is Nodeport and externalStartingPort is set to 0 then the broker IDs are not added and the nodeport port numbers will be chosen automatically by the K8s Service controller
	ExternalStartingPort int32 `json:"externalStartingPort"`
```
The two `+kubebuilder:validation` markers are already part of this PR.  
By the way, the reason, `0` is accepted is precisely because of that exception which I think comes from [this bit of code](https://github.com/banzaicloud/koperator/blob/master/pkg/resources/nodeportexternalaccess/service.go#L34-L37). Please correct me if I'm wrong on this.

Reply:
> 

@mihaialexandrescu it looks solid to me, on top of what you have, I would also add something like this:
```
the choice of broker ID and externalStartingPort must satisfy 0 < broker.ID + externalStartingPort <= 65535
```

Reply:
So this ?

```
// ExternalListenerConfig defines the external listener config for Kafka
type ExternalListenerConfig struct {
	... ...
	// +kubebuilder:validation:Minimum=0
	// +kubebuilder:validation:Maximum=65535
	// externalStartingPort is added to each broker ID to get the port number that will be used for external access to the broker. 
         // The choice of broker ID and externalStartingPort must satisfy 0 < brokers.id + externalStartingPort <= 65535
	// If accessMethod is Nodeport and externalStartingPort is set to 0 then the broker IDs are not added and the nodeport port numbers will be chosen automatically by the K8s Service controller
	ExternalStartingPort int32 `json:"externalStartingPort"`
```

Reply:
That LGTM

Reply:
Got an ok offline from Marton on the phrasing.

Done in d1064eb3c0b18e5d4bf3da9ea67f73b070316445 (`kafkacluster_types.go` and the CRD updates for the description)